lib/recursive-methods.rb

```diff
@@ -1,49 +1,74 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return raise ArgumentError if n < 0
+  return 1 if n == 0 || n == 1
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  reverse_s = reverse(s[1..-1])
+  reverse_s += s[0]
+  return reverse_s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def reverse_inplace(s)
```

Comment:
⚠️ This is not in place as with each iteration you're creating a new string.  

Consider adjusting the header to be like this.

`def reverse_inplace(s, low=0, high=s.length - 1)`